lib/id3c/cli/command/etl/__init__.py

```diff
@@ -378,17 +378,76 @@ def upsert_encounter_location(db: DatabaseSession,
     """
     Upserts an encounter location by its *encounter_id* and *relation*.
     """
-    LOG.debug(f"Upserting encounter {relation} location")
+    LOG.debug(f"Upserting encounter with location '{relation}'")
+
+    data = {
+        "encounter_id": encounter_id,
+        "relation": relation,
+        "location_id": location_id,
+    }
 
     with db.cursor() as cursor:
         cursor.execute("""
-            insert into warehouse.encounter_location (encounter_id, relation, location_id)
-                values (%s, %s, %s)
-                on conflict (encounter_id, relation) do update
-                    set location_id = excluded.location_id
-            """, (encounter_id, relation, location_id))
+            select
+                encounter_id,
+                relation,
+                encounter_location.location_id != %(location_id)s::integer
+                    as data_changed
+            from warehouse.encounter_location
+            where
+                encounter_id = %(encounter_id)s and
+                relation = %(relation)s
+            for update
+            """, data
+        )
+
+        encounter_locs = list(cursor)
+
+    if not encounter_locs:
+        LOG.info("Adding new encounter location")
+
+        encounter_loc = db.fetch_row("""
+            insert into warehouse.encounter_location(
+                    encounter_id,
+                    relation,
+                    location_id
+                )
+                values (
+                    %(encounter_id)s,
+                    %(relation)s,
+                    %(location_id)s
+                )
+            returning encounter_id, relation
+            """, data
+        )
+
+    elif len(encounter_locs) == 1:
+        encounter_loc = encounter_locs[0]
+
+        LOG.info("Updating existing encounter location")
+
+        if encounter_loc.data_changed == False:
+            LOG.info(f"Skipping upsert for encounter with location '{relation}' (no change).")
+            return encounter_loc
```

Comment:
Similar to the comment below, no need to return anything if we're matching the prior implementation.

Reply:
For sure on the matching prior implementation. Do you know why this upsert in particular has no return value? Seems like all the other upserts in this file do return the update they made.

Reply:
Yeah, I noticed that too. My guess is that the returned value isn't needed anywhere in the process, just confirmation that it was upserted.